src/modm/platform/comp/stm32/comp.hpp.in

```diff
@@ -14,701 +12,454 @@
-			getHysteresis()
-			{
-				return static_cast<Hysteresis>
-					(COMP{{ id }}->CSR & COMP{{ id }}_CSR_COMP{{ id }}HYST);
-			}
+		/**
+		 * \brief	Selects the hysteresis.
+		 */
+		static inline void
+		setHysteresis(Hysteresis hyst)
+		{
+			COMP{{ id }}->CSR = (COMP{{ id }}->CSR & ~HysteresisMask) | static_cast<uint32_t>(hyst);
+		}
 
-			/**
-			 * \brief	Selects the blanking source.
-			 */
-			static inline void
-			setBlankingSource(BlankingSource blanking)
-			{
-				COMP{{ id }}->CSR |=
-					(COMP{{ id }}->CSR & ~COMP{{ id }}_CSR_COMP{{ id }}BLANKING)
-					| static_cast<uint32_t>(blanking);
-			}
+		/**.
+		 * \brief	Returns the hysteresis.
+		 */
+		static inline Hysteresis
+		getHysteresis()
+		{
+			return static_cast<Hysteresis> (COMP{{ id }}->CSR & HysteresisMask);
+		}
 
-			/**.
-			 * \brief	Returns the blanking source.
-			 */
-			static inline BlankingSource
-			getBlankingSource()
-			{
-				return static_cast<BlankingSource>
-					(COMP{{ id }}->CSR & COMP{{ id }}_CSR_COMP{{ id }}BLANKING);
-			}
+		/**
+		 * \brief	Selects the blanking source.
+		 */
+		static inline void
+		setBlankingSource(BlankingSource blanking)
+		{
+			COMP{{ id }}->CSR |= (COMP{{ id }}->CSR & ~BlankingSourceMask) | static_cast<uint32_t>(blanking);
+		}
 
-			/**
-			 * \brief	Returns the current Comparator output.
-			 */
-			static inline bool
-			getOutput()
-			{
-				return COMP{{ id }}->CSR & COMP{{ id }}_CSR_COMP{{ id }}OUT;
-			}
+		/**.
+		 * \brief	Returns the blanking source.
+		 */
+		static inline BlankingSource
+		getBlankingSource()
+		{
+			return static_cast<BlankingSource>(COMP{{ id }}->CSR & BlankingSourceMask);
+		}
 
-			/**
-			 * \brief	Locks the comparator.
-			 *
-			 * Comparator can only be unlocked by a system reset.
-			 */
-			static inline void
-			lock(void)
-			{
-				COMP{{ id }}->CSR |= COMP{{ id }}_CSR_COMP{{ id }}LOCK;
-			}
+		/**
+		 * \brief	Returns the current Comparator output.
+		 */
+		static inline bool
+		getOutput()
+		{
+		{% if target.family == "f3" -%}
+			return COMP{{ id }}->CSR & {{ csr }}OUT;
+		{% elif target.family == "l4" -%}
+			return COMP{{ id }}->CSR & {{ csr }}VALUE;
+		{% endif -%}
+		}
 
-			/**
-			 * \brief	Returns true if the comparator is locked.
-			 */
-			static inline bool
-			isLocked(void)
-			{
-				return COMP{{ id }}->CSR & COMP{{ id }}_CSR_COMP{{ id }}LOCK;
-			}
-		};
-	}
+		/**
+		 * \brief	Locks the comparator.
+		 *
+		 * Comparator can only be unlocked by a system reset.
+		 */
+		static inline void
+		lock(void)
+		{
+			COMP{{ id }}->CSR |= {{ csr }}LOCK;
+		}
+
+		/**
+		 * \brief	Returns true if the comparator is locked.
+		 */
+		static inline bool
+		isLocked(void)
+		{
+			return COMP{{ id }}->CSR & {{ csr }}LOCK;
+		}
+
+	public:
+		// start inherited documentation
+		template< template<Peripheral _> class... Signals >
+		static void
+		connect()
+		{
+			using Connector = GpioConnector<Peripheral::Comp{{ id }}, Signals...>;
+			Connector::connect();
```

Comment:
FYI: Is the [current connect behaviour correct](https://github.com/modm-io/modm/blob/develop/src/modm/platform/gpio/stm32/pin.hpp.in#L204-L209
)? ie. no `setAnalogInput()`?
```cpp
	connect()
	{
		%% if signal.af | length
		setAlternateFunction({{ signal.af[0] }});
		%% elif signal.driver.startswith("Adc") or signal.driver.startswith("Dac")
		disconnect();
		setAnalogInput();
		%% endif
	}
```

Reply:
The reference manuals state:
> The I/Os used as comparators inputs must be configured in analog mode in the GPIOs registers.
> The comparator output can be connected to the I/Os using the alternate function channel given in “Alternate function mapping” table in the datasheet.

Sounds like we should do `setAnalogInput()` for the input pins.
But I can confirm the comparator inputs and outputs from the two examples are working correctly.

Reply:
Hm, this doesn't generate anything for Comp Inputs right now, since they do not have any AF number in the modm-devices dataset. You should add ` or signal.driver.startswith("Comp")` to the line to set analog mode for these too.

```xml
<signal driver="comp" instance="1" name="inp"/>
```

The Comp output has a AF, so it will set that instead.

```xml
<signal af="8" driver="comp" instance="2" name="out"/>
```

Reply:
Is this to verbose?
```python
if ... or (signal.driver.startswith("Comp") and signal.name.startswith("In")):
```